### PR TITLE
Access ViewStore state dynamically via closure

### PR DESCRIPTION
**Pull request by gordonbrander**
Update `ViewStore` so that it dynamically gets the state of the parent store using a closure. Previously we were copying state by value.

This PR updates ViewStore's state-getting approach to work the same way as Bindings. Bindings also use a `get` and `set` closure under the hood to implement bi-directional binding to a source of truth.

## Motivation and background

This is in response to an unusual bug we experienced while working on Subconscious.

Was experiencing a mysterious crasher when factoring NavigationStack into a subview. What I discovered was that, at some point, NavigationStack seems to be attempting to mutate the array that manages that stack. I believe the fact that we pass the stack down as a value type may be causing the problem. It's obscured because the issue happens in Apple's proprietary SwiftUI code. The crash had to do with an array manipulation, and seemed likely to be related to the array that manages the view stack, or perhaps it with some sort of race condition in view rendering on the SwiftUI side.

The workaround was giving NavigationStack a @State binding and replaying changes onto it.

However, this made me wonder if the cause was the fact that we passed the stack by value down the the view, before creating a Binding that referenced it.

After stubbing in a ViewStore with a Binding-like closure get function, the issue was resolved. So it seems that there are corner cases where SwiftUI needs to dynamically read the value via closure, not have it passed down by value.

### A note on SwiftUI's update strategy, and why this may be happening

> SwiftUI only tracks `@State` and `@ObservedObject` variables inside a View. More accurately, Any views that reads from `@State` var foo are invalidated whenever foo changes, wherever foo is declared, and Any views that has `@ObservedObject` var foo are invalidated whenever foo.objectWillChange is triggered. Binding is just a way to read from/write to variables across views. If you mutate to ObservableObject, it will trigger objectWillChange, but if the view that mutates it does not observe it with `@ObservedObject` the view will not be invalidated. https://forums.swift.org/t/swiftui-bindings-does-reading-them-in-body-mean-view-will-re-render/34600/2

I believe this description is correct, based on my experience with SwiftUI. So what a Binding does is providing a dynamic window into a slice of state, but Binding itself does not trigger view invalidation. Instead, mutation happens to the parent `@State`, `@StateObject`, or `@ObservedObject`, through the Binding's set closure. This invalidates the parent containing the `ObservableObject` or state, which causes its body to be re-evaluated, which will in-turn cause its children's body to be re-evaluated.

In theory, passing copies of state down, and messages up should accomplish the same kind of top-down rendering cascade. In practice, this almost always works just fine. Apple's SwiftUI tutorial shows passing down state to children by value, for example. However, in very rare edge cases, it seems to trip up. Some areas we've had to work around issues are inputs, and NavigationStack. In the case of passing down state to a child, the Apple tutorial still uses `@State` in the child view for state like inputs owned by that view... Why this should be a problem is a mystery that can probably only be explained by the closed-source internals of SwiftUI. Perhaps Apple may have some race condition around view invalidation, or something like that. Either way, using a closure to get state, the way Binding does, seems to side-step these issues.

What else does this tell us? Binding (and, by extension, ViewStore) do not optimize re-rendering. A child view with a binding will re-render any time anything in the root ObservableObject changes (Store, in our case). So we could achieve a similar effect by passing down the whole store. However, Binding and ViewStore do give us the advantage of type-erased domain-specific encapsulation.

## Other possible approaches

- There may be an opportunity to use EquatableView to customize the view diffing and make it more efficient. If we explore this path, we must take care to remember that the two instances of ObservedObject will have the same identity, so we would have to diff by another means, such as keeping the previous model state as a private property of the ViewStore.
- Another approach could be to create ObservableObjects for sub-components, and replay select events onto them from the parent store using Combine. This is the approach ComposableArchitecture takes. However, ObservableObjects must be classes, and this means re-creating a new instance of the heap-allocated class every time the parent store changes, plus any escaping closures that are needed to map the Combine publisher. I don't think this buys us much.
    - Note that ViewStore requires 2 escaping closures. These are heap-allocated too. We could alternatively use a protocol.

Pursuing the ViewStore approach because it is the most straightforward.

### CODE REVIEW

File: Sources/ObservableStore/ObservableStore.swift [hunk 1]
```diff
@@ -223,6 +223,18 @@ where Model: ModelProtocol
...
     }
 
+    /// Initialize and send an initial action to the store.
+    /// Useful when performing actions once and only once upon creation
+    /// of the store.
+    public convenience init(
+        state: Model,
+        action: Model.Action,
+        environment: Model.Environment
+    ) {
```
Comment (gordonbrander) [inline Sources/ObservableStore/ObservableStore.swift:233]: Unrelated change. Bringing this back to make it easier to upgrade ObservableStore in Subconscious.

There was really no good reason to remove it, except for parsimony, and we use it in a handful of places in Subconscious.

File: Sources/ObservableStore/ObservableStore.swift [hunk 2]
```diff
@@ -368,6 +403,139 @@ public struct Address {
     }
 }
 
+/// A cursor provides a complete description of how to map from one component
```
Comment (gordonbrander) [inline Sources/ObservableStore/ObservableStore.swift:406]: Unrelated change. Bringing back CursorProtocol and KeyedCursorProtocol with deprecation warnings.

We should have done this before, rather than removing them outright. I didn't know about deprecation flags at the time.

Motivation for bringing these back now is to make upgrading ObservableStore in Subconscious easier.

**Comment by gordonbrander** [comment]
Note that by bringing back these deprecated methods, I was able to drop-in this branch into Subconscious without problems.

**Comment by gordonbrander** [comment]
We should also investigate DynamicProperty (Binding conforms to this).

> When your property wrapper conforms to the DynamicProperty protocol, your property wrapper will essentially become a part of your SwiftUI view. This means that your property wrapper can extract values from your SwiftUI environment, and your SwiftUI view will ask your property wrapper to update itself whenever it's about to evaluate your view's body. You can even have `@StateObject` properties in your property wrapper to trigger updates in your views. https://www.donnywals.com/writing-custom-property-wrappers-for-swiftui/

That sounds promising.

Resources:

- https://developer.apple.com/documentation/swiftui/dynamicproperty
- https://www.hackingwithswift.com/plus/intermediate-swiftui/creating-a-custom-property-wrapper-using-dynamicproperty
- https://www.donnywals.com/writing-custom-property-wrappers-for-swiftui/
- https://davedelong.com/blog/2021/04/02/custom-property-wrappers-for-swiftui/

Related discussions in TCA:

- https://github.com/search?q=repo%3Apointfreeco%2Fswift-composable-architecture+dynamicproperty&type=discussions
- https://github.com/pointfreeco/swift-composable-architecture/pull/935
- https://github.com/pointfreeco/swift-composable-architecture/discussions/1209
- https://github.com/pointfreeco/swift-composable-architecture/discussions/1016

**Review by bfollington — APPROVED** [review]
Confusing, annoying, but LGTM as a fix. `DynamicProperty` definitely seems worth investigating.
